packages/vitest/src/node/execute.ts

```diff
@@ -38,23 +38,49 @@ export class VitestRunner extends ViteNodeRunner {
       },
     }
 
-    this.mocker = createMocker(this.root, options.mockMap)
+    this.mocker = new VitestMocker(this.root, options.mockMap)
   }
 
   prepareContext(context: Record<string, any>) {
     const request = context.__vite_ssr_import__
 
+    const resolvePath = async(id: string, importer: string) => {
+      const path = await this.options.resolveId(id, importer)
+      return {
+        path: normalizeId(path?.id || id),
+        external: path?.id.includes('/node_modules/') ? id : null,
+      }
+    }
+
+    const resolveMocks = async() => {
+      await Promise.all(this.mocker.pendingIds.map(async(mock) => {
+        const { path, external } = await resolvePath(mock.id, mock.importer)
+        if (mock.type === 'unmock')
+          this.mocker.unmockPath(path, external)
+        if (mock.type === 'mock')
+          this.mocker.mockPath(path, external, mock.factory)
+      }))
+
+      this.mocker.pendingIds = []
+    }
+
     const callFunctionMock = async(dep: string, mock: () => any) => {
       const cacheName = `${dep}__mock`
       const cached = this.moduleCache.get(cacheName)?.exports
       if (cached)
         return cached
+      this.mocker.processingDep = dep
```

Comment:
Does this work with concurrent import resolving? Like if you mock multiple modules at the same time, wouldn't them overrides each others?

Reply:
I was afraid it would do that, but I'm not ready to provide 100% stable solution.

Reply:
If you have any ideas, I'm ready to hear them. This was done to fix this: https://github.com/vitest-dev/vitest/issues/546

Reply:
Or we can just keep a set of traces and throw when circular imports are met as a temporary solution and improve later?

I am not so sure about the complexity added to solve it.

Reply:
Throwing would add more complexity :p

Reply:
I've just removed workaround for now.